### PR TITLE
feat: flexible length specifications

### PR DESCRIPTION
**Pull request by RonMcKay**
This PR implements more flexible lengths for periods and step sizes. With this, meta tags like the following are possible:

`[x] [period] [@ YYYY-MM[-DD]] [/ [m] [step]]`

See also the updated README.

Additionally I fixed an error when the start date has no day assigned. (1310bc7) And added some tests for the new feature. (9469c83)
To make the specification more fluid in terms of the language I also added the possibility to specify `days`, `weeks`, `months` and `years`.

Note for @Akuukis: After this PR I will have a look into the Bug #18.

### CODE REVIEW

**Comment by Akuukis** [comment]
This feature is good!

Sadly it has a bug, split produces too much transactions when used as a tag, see the new tests pushed to master.

**Comment by RonMcKay** [comment]
> This feature is good!
> 
> Sadly it has a bug, split produces too much transactions when used as a tag, see the new tests pushed to master.

I found the bug. When the mark contained only an integer the regex matched that to be the multiplier and used the default period (`month`) resulting in too many txn. I updated the regex with two negative look aheads for the end of the string and for an '/'. Now a single integer is matched as the `period` and the multiplier is by default set to 1. I rebased everything on master and the tests are passing. Can you very the fix on your ledger?

BTW: The tests are really handy for finding these type of bugs. :)

**Comment by Akuukis** [comment]
That one seems fixed :+1: But there's still some errors in my ledger. I isolated the next bug, see the new test on master. It seems related to #18, here are the options as I see it:
- let's adjust this PR so that it's broken in exactly same way (so that user's existing balances don't error)
- let's fix that first, and then return to this PR afterwards.
- maybe it's not related

Agree that tests are great. FYI I've already wrote at other [beancount_plugin_utils](https://github.com/Akuukis/beancount_plugin_utils) repo almost full coverage of beancount tests that *should* be ported to this repo, but my plan was to reuse that repo's exports across all my beancount projects, and test them there once instead. It would be a big breaking change, so I didn't do that in one go, but then blais announced a new version of beancount that would nullify most of my efforts in that library, so I've postponed that until the new beancount... Anyways, most of your work here I will want to port to beancount_plugin_utils someday.

**Comment by RonMcKay** [comment]
Okay that was not related to #18 but the test case helped anyways. I forgot to add a positive lookahead for 'a digit boundary' in the regex. When someone uses the `x` together with `period` there has to always be some space or a letter following `x`. Otherwise the number needs to be matched as the period itself.
This caused in the test case the `17` to be split between `x` and `period`, resulting in `x=1` and `period=7`.
This should be fixed now, thanks for investigating!

EDIT: I noticed that you already merged the update of the README to master, I guess that was not intended? But anyways, that should be unimportant as soon as the PR merges.

File: beancount_interpolate/common.py [hunk 1]
```diff
@@ -118,7 +123,7 @@ def parse_mark(mark, default_date, config):
 
 # Infer Duration, start and steps. Spacing optinonal. Format: [123|KEYWORD] [@ YYYY-MM[-DD]] [/ step]
```
Comment (Akuukis) [inline beancount_interpolate/common.py]: please update the comment here

File: beancount_interpolate/common.py [hunk 2]
```diff
@@ -118,7 +123,7 @@ def parse_mark(mark, default_date, config):
...
 # 0. max duration, 1. year, 2. month, 3. day, 4. min step
-RE_PARSING = re.compile(r"^\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([^-/\s]+)?\s*?)?$")
+RE_PARSING = re.compile(r"^\s*([0-9]+(?=[\sa-zA-Z]+)(?!\s*[@$/]))?\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([0-9]+)?\s*([^-/\s]+)?\s*?)?$")
```
Comment (Akuukis) [inline beancount_interpolate/common.py]: This regex was awful to start with, but now it's beyond comprehensive. Can you please document it? For example like this 
```
E_PARSING = re.compile(r"^\s*([0-9]+(?=[\sa-zA-Z]+)(?!\s*[@$/]))?\s*?([^-/\s]+)?\s*?(?:@\s*?([0-9]{4})-([0-9]{2})(?:-([0-9]{2}))?)?\s*?(?:\/\s*?([0-9]+)?\s*([^-/\s]+)?\s*?)?$")
#                            ( 1.                              )     ( 2.     )
# 1. the first thing doing it's thing
# 2. the second thing matching another thing
# 3. ...
```

I'd also like to take it to regex-specific testing, and am curious about white-space handling between the parts as well.

File: beancount_interpolate/common.py [hunk 3]
```diff
@@ -143,6 +148,7 @@ def distribute_over_period(params, default_date, total_value, config):
     amounts = []
     accumulated_remainder = D(str(0))
 
+    # TODO: In the following `begin_date + i * step` can result in an OverflowError
```
Comment (Akuukis) [inline beancount_interpolate/common.py:179]: please elaborate. Do I understand it correctly?
> in case `begin_date + duration` is so close to overflow, that last `begin_date + i * step` which would be beyond it and thus break the while loop raises OverflowError instead.

**Comment by Akuukis** [comment]
Thanks you, now PR runs on my ledger without errors and with the same performance as before :heavy_check_mark: 

Thanks for pointing out the README, I rebased and force pushed to correct that. :heavy_check_mark: 

See few comments inline mostly about documentation. :spiral_notepad:

File: beancount_interpolate/common.py (hunk 2 above)
Comment (RonMcKay) [reply beancount_interpolate/common.py]: Yes I can do that. I would then like to split the regex into multiple lines surrounded by braces.

Comment (RonMcKay) [reply beancount_interpolate/common.py]: This is then way easier to annotate

Comment (RonMcKay) [reply beancount_interpolate/common.py]: I also noticed that are some unnecessary `?` and maybe some other stuff.

Comment (RonMcKay) [reply beancount_interpolate/common.py]: A small test suite for different meta tags would be indeed quite nice.

File: beancount_interpolate/common.py (hunk 3 above)
Comment (RonMcKay) [reply beancount_interpolate/common.py:179]: Yes, this would be a corner case. But theoretically one could also specify a step size of `inf` or something like `1 year / 10000 years` which would lead to an OverflowError. The second case is probably not that common so I think the bigger problem will be the more 'hidden' `OverflowError` when 
`begin_date + i * step < end_date` and `begin_date + i * step <= today_date`
but
`begin_date + (i + 1) * step > MAX_DATE`
This would then lead to an `OverflowError` in 
`tmp_date = begin_date + i * step`

This is however not a problem of the new feature as before there was also the possibility of overflowing the `datetime` objects.

Comment (RonMcKay) [reply beancount_interpolate/common.py:179]: I will document the TODO in the source code a little more in detail for future reference.

**Comment by RonMcKay** [comment]
@Akuukis I updated the code documentation and tried to add some description of the regex expression. I also added some further comments on the possible `OverflowError`.

**Comment by RonMcKay** [comment]
Also just removed some unnecessary `?` from the regex expression.

EDIT: Just tell me if you want me to rebase on master for merging.

**Comment by Akuukis** [comment]
Thanks!
